File: lib/fsharp/dice_fsharp.rs
#![allow(dead_code)]
#![allow(non_camel_case_types)]
#![allow(non_snake_case)]
#![allow(non_upper_case_globals)]
#![allow(unreachable_code)]
#![allow(unused_attributes)]
#![allow(unused_imports)]
#![allow(unused_macros)]
#![allow(unused_parens)]
#![allow(unused_variables)]
#![allow(unused_assignments)]
use fable_library_rust::NativeArray_::array_from;
use fable_library_rust::String_::fromString;
mod module_fc267fc3 {
    pub mod Dice {
        use super::*;
        pub mod DiceFSharp {
            use super::*;
            use crate::Polyglot::Common::TraceLevel;
            use crate::Polyglot::Common::trace;
            use fable_library_rust::List_::List;
            use fable_library_rust::List_::cons;
            use fable_library_rust::List_::empty;
            use fable_library_rust::List_::head;
            use fable_library_rust::List_::initialize;
            use fable_library_rust::List_::isEmpty;
            use fable_library_rust::List_::length;
            use fable_library_rust::List_::tail;
            use fable_library_rust::List_::tryItem;
            use fable_library_rust::Native_::Func0;
            use fable_library_rust::Native_::Func1;
            use fable_library_rust::Native_::Func2;
            use fable_library_rust::Native_::LrcPtr;
            use fable_library_rust::Native_::MutCell;
            use fable_library_rust::Native_::getZero;
            use fable_library_rust::NativeArray_::Array;
            use fable_library_rust::Option_::iterate;
            use fable_library_rust::Seq_::cache;
            use fable_library_rust::Seq_::item as item_1;
            use fable_library_rust::Seq_::map;
            use fable_library_rust::Seq_::unfold;
            use fable_library_rust::String_::printfn;
            use fable_library_rust::String_::sprintf;
            use fable_library_rust::String_::string;
            use fable_library_rust::System::Collections::Generic::IEnumerable_1;
            ///
            /// ## sixthPowerSequence
            pub fn sixthPowerSequence() -> LrcPtr<dyn IEnumerable_1<i32>> {
                cache(unfold(
                    Func1::new(move |state: i32| Some(LrcPtr::new((state, (state) * 6_i32)))),
                    1_i32,
                ))
            }
            ///
            /// ## accumulateDiceRolls
            pub fn accumulateDiceRolls(
                log: Option<Func1<string, ()>>,
                rolls: List<i32>,
                power: i32,
                acc: i32,
            ) -> Option<LrcPtr<(i32, List<i32>)>> {
                let log = MutCell::new(log.clone());
                let rolls: MutCell<List<i32>> = MutCell::new(rolls.clone());
                let power: MutCell<i32> = MutCell::new(power);
                let acc: MutCell<i32> = MutCell::new(acc);
                '_accumulateDiceRolls: loop {
                    break '_accumulateDiceRolls (if (power.get().clone()) < 0_i32 {
                        iterate(
                            {
                                let arg: string = sprintf!(
                                    "accumulateDiceRolls / power: {} / acc: {}",
                                    power.get().clone(),
                                    acc.get().clone()
                                );
                                Func1::new({
                                    let arg = arg.clone();
                                    move |func: Func1<string, ()>| func(arg.clone())
                                })
                            },
                            log.get().clone(),
                        );
                        Some(LrcPtr::new((
                            (acc.get().clone()) + 1_i32,
                            rolls.get().clone(),
                        )))
                    } else {
                        if !isEmpty(rolls.get().clone()) {
                            if (head(rolls.get().clone())) > 1_i32 {
                                let rest_1: List<i32> = tail(rolls.get().clone());
                                let roll_1: i32 = head(rolls.get().clone());
                                let value: i32 = ((roll_1) - 1_i32)
                                    * (item_1(
                                        power.get().clone(),
                                        Dice::DiceFSharp::sixthPowerSequence(),
                                    ));
                                iterate(
                                    {
                                        let arg_1: string = sprintf!(
                                            "accumulateDiceRolls / power: {} / acc: {} / roll: {} / value: {}",
                                            power.get().clone(),
                                            acc.get().clone(),
                                            roll_1,
                                            value
                                        );
                                        Func1::new({
                                            let arg_1 = arg_1.clone();
                                            move |func_1: Func1<string, ()>| func_1(arg_1.clone())
                                        })
                                    },
                                    log.get().clone(),
                                );
                                {
                                    let log_temp = log.get().clone();
                                    let rolls_temp: List<i32> = rest_1.clone();
                                    let power_temp: i32 = (power.get().clone()) - 1_i32;
                                    let acc_temp: i32 = (acc.get().clone()) + (value);
                                    log.set(log_temp);
                                    rolls.set(rolls_temp);
                                    power.set(power_temp);
                                    acc.set(acc_temp);
                                    continue '_accumulateDiceRolls;
                                }
                            } else {
                                let rest_2: List<i32> = tail(rolls.get().clone());
                                let roll_2: i32 = head(rolls.get().clone());
                                iterate(
                                    {
                                        let arg_2: string = sprintf!(
                                            "accumulateDiceRolls / power: {} / acc: {} / roll: {}",
                                            power.get().clone(),
                                            acc.get().clone(),
                                            roll_2
                                        );
                                        Func1::new({
                                            let arg_2 = arg_2.clone();
                                            move |func_2: Func1<string, ()>| func_2(arg_2.clone())
                                        })
                                    },
                                    log.get().clone(),
                                );
                                {
                                    let log_temp = log.get().clone();
                                    let rolls_temp: List<i32> = rest_2.clone();
                                    let power_temp: i32 = (power.get().clone()) - 1_i32;
                                    let acc_temp: i32 = acc.get().clone();
                                    log.set(log_temp);
                                    rolls.set(rolls_temp);
                                    power.set(power_temp);
                                    acc.set(acc_temp);
                                    continue '_accumulateDiceRolls;
                                }
                            }
                        } else {
                            None::<LrcPtr<(i32, List<i32>)>>
                        }
                    });
                }
            }
            ///
            /// ## rollWithinBounds
            pub fn rollWithinBounds(
                log: Option<Func1<string, ()>>,
                max: i32,
                rolls: List<i32>,
            ) -> Option<i32> {
                let matchValue: Option<LrcPtr<(i32, List<i32>)>> =
                    Dice::DiceFSharp::accumulateDiceRolls(
                        log,
                        rolls.clone(),
                        (length(rolls)) - 1_i32,
                        0_i32,
                    );
                if matchValue.is_some() && {
                    let result: i32 = (matchValue.clone().unwrap()).0.clone();
                    if (result) >= 1_i32 {
                        (result) <= (max)
                    } else {
                        false
                    }
                } {
                    let result_1: i32 = (matchValue.clone().unwrap()).0.clone();
                    Some(result_1)
                } else {
                    None::<i32>
                }
            }
            ///
            /// ## rollDice
            pub fn rollDice() -> i32 {
                rand::Rng::gen_range(&mut rand::thread_rng(), 1..7)
            }
            ///
            /// ## rotateNumber
            pub fn rotateNumber(max: i32, n: i32) -> i32 {
                ((((n) - 1_i32) + (max)) % (max)) + 1_i32
            }
            ///
            /// ## rotateNumbers
            pub fn rotateNumbers(
                max: i32,
                items: LrcPtr<dyn IEnumerable_1<i32>>,
            ) -> LrcPtr<dyn IEnumerable_1<i32>> {
                map(
                    Func1::new({
                        let max = max.clone();
                        move |n: i32| Dice::DiceFSharp::rotateNumber(max, n)
                    }),
                    items,
                )
            }
            ///
            /// ## createSequentialRoller
            pub fn createSequentialRoller<a: Clone + 'static>(list: List<a>) -> Func0<a> {
                let currentIndex: LrcPtr<MutCell<i32>> = LrcPtr::new(MutCell::new(0_i32));
                Func0::new({
                    let currentIndex = currentIndex.clone();
                    let list = list.clone();
                    move || {
                        let matchValue: Option<a> =
                            tryItem(currentIndex.get().clone(), list.clone());
                        match &matchValue {
                            None => panic!("{}", string("createSequentialRoller / End of list"),),
                            Some(matchValue_0_0) => {
                                let item: a = matchValue_0_0.clone();
                                currentIndex.set((currentIndex.get().clone()) + 1_i32);
                                item
                            }
                        }
                    }
                })
            }
            ///
            /// ## rollProgressively
            pub fn rollProgressively(
                log: Option<Func1<string, ()>>,
                roll: Func0<i32>,
                reroll: bool,
                max: i32,
            ) -> i32 {
                let power: i32 = ({
                    let max_1: i32 = max;
                    let 루프 = Func2::new({
                        let log = log.clone();
                        let max_1 = max_1.clone();
                        move |n: i32, p: i32| {
                            let n: MutCell<i32> = MutCell::new(n);
                            let p: MutCell<i32> = MutCell::new(p);
                            '_루프: loop {
                                break '_루프 (if (p.get().clone()) < (max_1) {
                                    let n_temp: i32 = (n.get().clone()) + 1_i32;
                                    let p_temp: i32 = (p.get().clone()) * 6_i32;
                                    n.set(n_temp);
                                    p.set(p_temp);
                                    continue '_루프;
                                } else {
                                    iterate(
                                        {
                                            let arg: string = sprintf!(
                                                "calculateDiceCount / max: {} / n: {} / p: {}",
                                                max_1,
                                                n.get().clone(),
                                                p.get().clone()
                                            );
                                            Func1::new({
                                                let arg = arg.clone();
                                                move |func: Func1<string, ()>| func(arg.clone())
                                            })
                                        },
                                        log.clone(),
                                    );
                                    n.get().clone()
                                });
                            }
                        }
                    });
                    if (max_1) == 1_i32 {
                        1_i32
                    } else {
                        루프(0_i32, 1_i32)
                    }
                }) - 1_i32;
                let 루프_1 = Func2::new({
                    let log = log.clone();
                    let max = max.clone();
                    let power = power.clone();
                    let reroll = reroll.clone();
                    let roll = roll.clone();
                    move |rolls: List<i32>, size: i32| {
                        let rolls: MutCell<List<i32>> = MutCell::new(rolls.clone());
                        let size: MutCell<i32> = MutCell::new(size);
                        '_루프_1: loop {
                            break '_루프_1 (if (size.get().clone()) < ((power) + 1_i32) {
                                let rolls_temp: List<i32> = cons(roll(), rolls.get().clone());
                                let size_temp: i32 = (size.get().clone()) + 1_i32;
                                rolls.set(rolls_temp);
                                size.set(size_temp);
                                continue '_루프_1;
                            } else {
                                let matchValue: Option<LrcPtr<(i32, List<i32>)>> =
                                    Dice::DiceFSharp::accumulateDiceRolls(
                                        log.clone(),
                                        rolls.get().clone(),
                                        power,
                                        0_i32,
                                    );
                                {
                                    let mut matchResult: i32 = getZero::<i32>();
                                    let mut result_1: i32 = getZero::<i32>();
                                    if matchValue.is_some() {
                                        if ((matchValue.clone().unwrap()).0.clone()) <= (max) {
                                            matchResult = 0_i32;
                                            result_1 = (matchValue.clone().unwrap()).0.clone();
                                        } else {
                                            if reroll {
                                                matchResult = 1_i32;
                                            } else {
                                                matchResult = 2_i32;
                                            }
                                        }
                                    } else {
                                        if reroll {
                                            matchResult = 1_i32;
                                        } else {
                                            matchResult = 2_i32;
                                        }
                                    };
                                    match &matchResult {
                                        0_i32 => result_1.clone(),
                                        1_i32 => {
                                            let rolls_temp: List<i32> = initialize(
                                                power,
                                                Func1::new({
                                                    let roll = roll.clone();
                                                    move |_arg: i32| roll()
                                                }),
                                            );
                                            let size_temp: i32 = power;
                                            rolls.set(rolls_temp);
                                            size.set(size_temp);
                                            continue '_루프_1;
                                        }
                                        2_i32 => {
                                            let rolls_temp: List<i32> =
                                                cons(roll(), rolls.get().clone());
                                            let size_temp: i32 = (size.get().clone()) + 1_i32;
                                            rolls.set(rolls_temp);
                                            size.set(size_temp);
                                            continue '_루프_1;
                                        }
                                        _ => unreachable!(),
                                    }
                                }
                            });
                        }
                    }
                });
                루프_1(empty::<i32>(), 0_i32)
            }
            ///
            /// ## main
            pub fn main(args: Array<string>) -> i32 {
                let result: i32 = Dice::DiceFSharp::rollProgressively(
                    Some({
                        fn clo(a0: string) {
                            printfn!("{}", a0);
                        }
                        Func1::new({
                            let clo = clo.clone();
                            move |arg: string| clo(arg)
                        })
                    }),
                    Func0::new(move || Dice::DiceFSharp::rollDice()),
                    true,
                    i32::MAX / 10_i32,
                );
                trace(
                    LrcPtr::new(TraceLevel::Debug),
                    Func0::new({
                        let result = result.clone();
                        move || sprintf!("main / result: {}", result)
                    }),
                    Func0::new(move || string("")),
                );
                0_i32
            }
        }
    }
}
pub use module_fc267fc3::*;
#[path = "../../deps/polyglot/deps/spiral/lib/spiral/async_.rs"]
mod module_763b9e6c;
pub use module_763b9e6c::*;
#[path = "../../deps/polyglot/deps/spiral/lib/spiral/common.rs"]
mod module_18f1c018;
pub use module_18f1c018::*;
#[path = "../../deps/polyglot/deps/spiral/lib/spiral/crypto.rs"]
mod module_c0301e36;
pub use module_c0301e36::*;
#[path = "../../deps/polyglot/deps/spiral/lib/spiral/date_time.rs"]
mod module_e5abed0b;
pub use module_e5abed0b::*;
#[path = "../../deps/polyglot/deps/spiral/lib/spiral/file_system.rs"]
mod module_336d1dc9;
pub use module_336d1dc9::*;
#[path = "../../deps/polyglot/deps/spiral/lib/spiral/lib.rs"]
mod module_12dde172;
pub use module_12dde172::*;
#[path = "../../deps/polyglot/deps/spiral/lib/spiral/networking.rs"]
mod module_7e35832b;
pub use module_7e35832b::*;
#[path = "../../deps/polyglot/deps/spiral/lib/spiral/platform.rs"]
mod module_afd02cca;
pub use module_afd02cca::*;
#[path = "../../deps/polyglot/deps/spiral/lib/spiral/runtime.rs"]
mod module_a3c64389;
pub use module_a3c64389::*;
#[path = "../../deps/polyglot/deps/spiral/lib/spiral/sm.rs"]
mod module_8f227ecb;
pub use module_8f227ecb::*;
#[path = "../../deps/polyglot/deps/spiral/lib/spiral/threading.rs"]
mod module_e08c16fb;
pub use module_e08c16fb::*;
#[path = "../../deps/polyglot/deps/spiral/lib/spiral/trace.rs"]
mod module_702ddcb4;
pub use module_702ddcb4::*;
#[path = "../../deps/polyglot/lib/fsharp/Common.rs"]
mod module_eb7e4788;
pub use module_eb7e4788::*;
pub mod Dice {
    pub use crate::module_fc267fc3::Dice::*;
}
pub mod Polyglot {
    pub use crate::module_eb7e4788::Polyglot::*;
}
pub fn main() {
    let args = std::env::args().skip(1).map(fromString).collect();
    Dice::DiceFSharp::main(array_from(args));
}

File: lib/fsharp/dice_fsharp.py
from __future__ import annotations
from fable_modules.fable_library.list import (is_empty, head, tail, FSharpList, length, try_item, cons, initialize, empty)
from fable_modules.fable_library.option import (to_array, value as value_1)
from fable_modules.fable_library.seq import (cache, unfold, iterate, item as item_1, map)
from fable_modules.fable_library.string_ import (to_console, printf)
from fable_modules.fable_library.types import Array
from fable_modules.fable_library.util import (IEnumerable_1, randint)
from collections.abc import Callable
import sys
from typing import (Any, TypeVar)
from lib.fsharp.common import (trace, TraceLevel)

__A = TypeVar("__A")

def sixth_power_sequence(__unit: None=None) -> IEnumerable_1[int]:
    """## sixthPowerSequence
    """
    def generator(state: int) -> tuple[int, int] | None:
        return (state, state * 6)

    return cache(unfold(generator, 1))


def accumulate_dice_rolls(log_mut: Callable[[str], None] | None, rolls_mut: FSharpList[int], power_mut: int, acc_mut: int) -> tuple[int, FSharpList[int]] | None:
    """## accumulateDiceRolls
    """
    while True:
        (log, rolls, power, acc) = (log_mut, rolls_mut, power_mut, acc_mut)
        if power < 0:
            def action(func: Callable[[str], None], log: Any=log, rolls: Any=rolls, power: Any=power, acc: Any=acc) -> None:
                func(((("accumulateDiceRolls / power: " + str(power)) + " / acc: ") + str(acc)) + "")

            iterate(action, to_array(log))
            return (acc + 1, rolls)

        elif not is_empty(rolls):
            if head(rolls) > 1:
                value: int = ((head(rolls) - 1) * item_1(power, sixth_power_sequence())) or 0
                def action_1(func_1: Callable[[str], None], log: Any=log, rolls: Any=rolls, power: Any=power, acc: Any=acc) -> None:
                    func_1(((((((("accumulateDiceRolls / power: " + str(power)) + " / acc: ") + str(acc)) + " / roll: ") + str(head(rolls))) + " / value: ") + str(value)) + "")

                iterate(action_1, to_array(log))
                log_mut = log
                rolls_mut = tail(rolls)
                power_mut = power - 1
                acc_mut = acc + value
                continue

            else: 
                def action_2(func_2: Callable[[str], None], log: Any=log, rolls: Any=rolls, power: Any=power, acc: Any=acc) -> None:
                    func_2(((((("accumulateDiceRolls / power: " + str(power)) + " / acc: ") + str(acc)) + " / roll: ") + str(head(rolls))) + "")

                iterate(action_2, to_array(log))
                log_mut = log
                rolls_mut = tail(rolls)
                power_mut = power - 1
                acc_mut = acc
                continue


        else: 
            return None

        break


def roll_within_bounds(log: Callable[[str], None] | None, max: int, rolls: FSharpList[int]) -> int | None:
    """## rollWithinBounds
    """
    match_value: tuple[int, FSharpList[int]] | None = accumulate_dice_rolls(log, rolls, length(rolls) - 1, 0)
    (pattern_matching_result, result_1) = (None, None)
    if match_value is not None:
        def _arrow216(__unit: None=None, log: Any=log, max: Any=max, rolls: Any=rolls) -> bool:
            result: int = match_value[0] or 0
            return (result <= max) if (result >= 1) else False

        if _arrow216():
            pattern_matching_result = 0
            result_1 = match_value[0]

        else: 
            pattern_matching_result = 1


    else: 
        pattern_matching_result = 1

    if pattern_matching_result == 0:
        return result_1

    elif pattern_matching_result == 1:
        return None



random: Any = {}

def roll_dice(__unit: None=None) -> int:
    return randint(1, 7)


def rotate_number(max: int, n: int) -> int:
    """## rotateNumber
    """
    return (((n - 1) + max) % max) + 1


def rotate_numbers(max: int, items: IEnumerable_1[int]) -> IEnumerable_1[int]:
    """## rotateNumbers
    """
    def mapping(n: int, max: Any=max, items: Any=items) -> int:
        return rotate_number(max, n)

    return map(mapping, items)


def create_sequential_roller(list_1: FSharpList[Any]) -> Callable[[], __A]:
    """## createSequentialRoller
    """
    current_index: int = 0
    def _arrow217(__unit: None=None, list_1: Any=list_1) -> __A:
        nonlocal current_index
        match_value: __A | None = try_item(current_index, list_1)
        if match_value is None:
            raise Exception("createSequentialRoller / End of list")

        else: 
            item: __A = value_1(match_value)
            current_index = (current_index + 1) or 0
            return item


    return _arrow217


def roll_progressively(log: Callable[[str], None] | None, roll: Callable[[], int], reroll: bool, max: int) -> int:
    """## rollProgressively
    """
    def _arrow218(__unit: None=None, log: Any=log, roll: Any=roll, reroll: Any=reroll, max: Any=max) -> int:
        max_1: int = max or 0
        def 루프(n_mut: int, p_mut: int) -> int:
            while True:
                (n, p) = (n_mut, p_mut)
                if p < max_1:
                    n_mut = n + 1
                    p_mut = p * 6
                    continue

                else: 
                    def action(func: Callable[[str], None], n: Any=n, p: Any=p) -> None:
                        func(((((("calculateDiceCount / max: " + str(max_1)) + " / n: ") + str(n)) + " / p: ") + str(p)) + "")

                    iterate(action, to_array(log))
                    return n

                break

        return 1 if (max_1 == 1) else 루프(0, 1)

    power: int = (_arrow218() - 1) or 0
    def 루프_1(rolls_mut: FSharpList[int], size_mut: int, log: Any=log, roll: Any=roll, reroll: Any=reroll, max: Any=max) -> int:
        while True:
            (rolls, size) = (rolls_mut, size_mut)
            if size < (power + 1):
                rolls_mut = cons(roll(None), rolls)
                size_mut = size + 1
                continue

            else: 
                match_value: tuple[int, FSharpList[int]] | None = accumulate_dice_rolls(log, rolls, power, 0)
                (pattern_matching_result, result_1) = (None, None)
                if match_value is not None:
                    if match_value[0] <= max:
                        pattern_matching_result = 0
                        result_1 = match_value[0]

                    elif reroll:
                        pattern_matching_result = 1

                    else: 
                        pattern_matching_result = 2


                elif reroll:
                    pattern_matching_result = 1

                else: 
                    pattern_matching_result = 2

                if pattern_matching_result == 0:
                    return result_1

                elif pattern_matching_result == 1:
                    def _arrow219(_arg: int, rolls: Any=rolls, size: Any=size) -> int:
                        return roll(None)

                    rolls_mut = initialize(power, _arrow219)
                    size_mut = power
                    continue

                elif pattern_matching_result == 2:
                    rolls_mut = cons(roll(None), rolls)
                    size_mut = size + 1
                    continue


            break

    return 루프_1(empty(), 0)


def main(args: Array[str]) -> int:
    """## main
    """
    def _arrow221(__unit: None=None, args: Any=args) -> Callable[[str], None]:
        clo: Callable[[str], None] = to_console(printf("%s"))
        def _arrow220(arg: str) -> None:
            clo(arg)

        return _arrow220

    def _arrow222(__unit: None=None, args: Any=args) -> int:
        return roll_dice()

    result: int = roll_progressively(_arrow221(), _arrow222, True, 2147483647 // 10) or 0
    def _arrow223(__unit: None=None, args: Any=args) -> str:
        return ("main / result: " + str(result)) + ""

    def _arrow224(__unit: None=None, args: Any=args) -> str:
        return ""

    trace(TraceLevel(1), _arrow223, _arrow224)
    return 0


if __name__ == "__main__":
    main(sys.argv[1:])


